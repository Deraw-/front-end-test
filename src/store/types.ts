// tslint:disable-next-line:no-empty-interface
export interface RootState {}

export interface Product extends Object {
	id: number;
	quantity: number;
}

export interface CartProduct extends Product {
	name: string;
	price: number;
	image: string;
}

// This is used to properly type components using mappers from Vuex
// see https://github.com/vuejs/vuex/pull/1121
// tslint:disable-next-line: no-empty-interface
export default interface VuexBindings {
	cartProducts: CartProduct[];
}
