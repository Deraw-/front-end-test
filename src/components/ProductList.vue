<template>
	<div>
		<h2 class="title">
			{{ $t('components.productList.title') }}
		</h2>

		<p
			class="description"
			v-html="$t('components.productList.description')"
		/>

		<ul class="products">
			<Product
				v-for="product in products"
				:key="product.id"
				:product="product"
			/>
		</ul>
	</div>
</template>

<script lang="ts">
	import Vue from 'vue';

	import Product from '@/components/Product.vue';

	export default Vue.extend({
		name: 'ProductList',
		components: {
			Product
		},
		props: {
			products: {
				type: [Object, Array],
				required: true
			}
		}
	});
</script>

<style lang="scss" scoped>
	.products {
		display: grid;
		grid-gap: 32px;
		grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
		grid-auto-rows: minmax(250px, auto);
		margin-top: 50px;
	}

	@media only screen and (max-width: 600px) {
		.products {
			margin-top: 25px;
		}
	}
</style>
