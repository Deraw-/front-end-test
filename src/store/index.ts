import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import VuexPersistence from 'vuex-persist';
import { RootState } from './types';

// Modules
import cart from './modules/cart';
import products from './modules/products';

// The store is saved in the browser's session
const vuexLocal = new VuexPersistence({
	storage: window.sessionStorage
});

Vue.use(Vuex);

// See https://vuex.vuejs.org/fr/getting-started.html for help
const store: StoreOptions<RootState> = {
	strict: true,
	state: {},
	// See https://vuex.vuejs.org/guide/modules.html for more info on modules
	modules: {
		cart,
		products
	},
	plugins: [
		vuexLocal.plugin
	]
};

export default new Vuex.Store<RootState>(store);
