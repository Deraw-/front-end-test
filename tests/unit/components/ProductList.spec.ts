import { shallowMount } from '@vue/test-utils';
import ProductList from '@/components/ProductList.vue';

import localVue from 'tests-unit/index';
import store from '@/store/';

// Data
import { CartProduct } from '@/store/types';
const products: CartProduct[] = [];

for (let index = 0; index < 8; index++) {
	products.push({
		id: index,
		quantity: 15,
		name: 'name',
		price: 15,
		image: '/img.jpg'
	});
}

// Build component before each test
let wrapper: any;
beforeEach(() => {
	wrapper = shallowMount(ProductList, {
		localVue,
		store,
		propsData: {
			products
		}
	});
});

// Tests
describe('ProductList.vue', () => {
	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
