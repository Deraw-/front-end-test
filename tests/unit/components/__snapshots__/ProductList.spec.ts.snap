// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList.vue renders correctly 1`] = `
<div>
  <h2 class="title">
    components.productList.title
  </h2>
  <p class="description">components.productList.description</p>
  <ul class="grid-list">
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
    <product-stub product="[object Object]"></product-stub>
  </ul>
</div>
`;
